1.Population dynamics

In [6]:
import numpy as np

A = [[0.70,0.15,0.10],
     [0.15,0.80,0.30],
     [0.15,0.05,0.60]]

x0 = [300,350,200]
# numpy's dot() function multiplys two matrices, 
# and gives an error if the sizes dont match up
print(np.dot(A,x0))

[282.5 385.  182.5]


a)  Use Eq. (1) to find the population distribution one year after the census.

In [8]:
x1 = np.dot(A,x0)
print(x1)

[282.5 385.  182.5]


b) Give a formula for Xn in terms of powers of A and xo.

xn = A*xn-1

c) Calculate the state vectors x1 , x2, ••. , x10. Observe that the population distribution
seems to be reaching a steady state. Estimate the steady-state population for
each region.

In [16]:
xn = np.dot(A,x0)
print("x 0 : ", x0)
print("x 1 : ", xn)

for i in range (9):
    xn = np.dot(A,xn)
    print("x",i+2,": ", xn)

x 0 :  [300, 350, 200]
x 1 :  [282.5 385.  182.5]
x 2 :  [273.75  405.125 171.125]
x 3 :  [269.50625 416.5     163.99375]
x 4 :  [267.52875   422.8240625 159.6471875]
x 5 :  [266.65845312 426.28271875 157.05882812]
x 6 :  [266.30920781 428.14259141 155.54820078]
x 7 :  [266.19265426 429.12491453 154.68243121]
x 8 :  [266.17183828 429.63355913 154.19460259]
x 9 :  [266.18478093 429.89100382 153.92421525]
x 10 :  [266.20541875 430.01778477 153.77679648]


It seems that

A. approaches 266\
B. approaches 430\
C. approaches 154

d) Calculate x20 and compare it with your estimate from part c).

In [18]:
xn = np.dot(A,x0)

for i in range (19):
    xn = np.dot(A,xn)
print("x",i+2,": ", xn)

x 20 :  [266.26471079 430.12092489 153.61436432]


I guess everything drifted less than I expected.

e) Let x_ 1 denote the state vector one year prior to the census. Calculate x_ 1 .

x_0 = Ax_1

x_1 = $A^{-1}$x_0

In [24]:
#numpy's way of taking an inverse
An1 = np.linalg.inv(A)

x_1 = np.dot(An1,x0)
print(x_1)
print("Taking the next step to check \n", np.dot(A,x_1))

[334.14634146 290.24390244 225.6097561 ]
Taking the next step to check 
 [300. 350. 200.]


f) Demonstrate that Eq. (1) has not always been an accurate model for population
distribution by calculating the state vector four years prior to the census.

In [31]:
print("x_ 1 : ",x_1)
x_n = np.dot(An1,x0)
for i in range(3):
    x_n = np.dot(An1,x_n)
    print("x_",i+2,": ", x_n)

x_ 1 :  [334.14634146 290.24390244 225.6097561 ]
x_ 2 :  [399.38528654 190.30339084 260.31132263]
x_ 3 :  [521.71536171  27.12139019 301.1632481 ]
x_ 4 :  [ 747.15971169 -231.60930611  334.44959442]


Region B ends up having a negative population... Which is not possible

g) How should we rearrange the population just after the census so that the distribution
three years later is x3 = [250, 400, 200]t? That is, what should x0 be in order to hit
the target X3?

In [44]:
x3 = [250,400,200]

print("x_ 3 : ",x3)
x_n = np.dot(An1,x3)
for i in range(3):
    x_n = np.dot(An1,x_n)
    print("x_",2-i,": ", x_n)

x_ 3 :  [250, 400, 200]
x_ 2 :  [242.07151828 289.45072378 318.47775795]
x_ 1 :  [249.79909902 143.84117726 456.35972372]
x_ 0 :  [ 286.04061999 -136.50545949  700.46483949]


In [46]:
#notice we have to start with a negitive population
year0 = [ 286.04061999, -136.50545949, 700.46483949]
xn = np.dot(A,year0)
print("start year 0 : ", year0)
# print("x 1 : ", xn)

for i in range (3):
    xn = np.dot(A,xn)
    print("year",i+1,": ", xn)

start year 0 :  [286.04061999, -136.50545949, 700.46483949]
year 1 :  [242.07151827 289.45072377 318.47775795]
year 2 :  [244.71544715 363.41463414 241.8699187 ]
year 3 :  [250. 400. 200.]
